Update dependency list

andyleejordan opened the pull request:
I've updated the dependency lists in `download.sh`, `New-UnixPackage`, and the Linux installation documentation to match those used in the Dockerfiles.

I feel that this probably needs to be pruned; but I am unsure which dependencies we do not need to specify (probably ones listed in "required" for the base distributions, and possibly some that .NET just states as required but is not (like libgcc?)). I would really like the dependency list to be reviewed by others.

This resolves #1937.

Also resolves #2211.

andyleejordan committed:
Add all dependencies to New-UnixPackage

Now that I finally have a decent list of dependencies (though it may
still be too many), I have expanded the dependencies of the package so
that users get better error messages.

I've removed ca-certificates as that's not *necessarily* required (but
needed on Docker to download over HTTPS). This may be true of other
packages too.

andyleejordan committed:
Update download.sh to install all dependencies

andyleejordan committed:
Update Linux installation docs to resolve dependencies

Use `dpkg -i` followed by `apt-get install -f`. Alternatives considered
included using `apt install`, which does not work, and copying the
PowerShell package to `/var/cache/apt/archives` which is just as messy
as this. At least this follows the same steps as `download.sh`.

msftclas commented:
Hi __@andschwa__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Andy Schwartzmeyer). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

vors commented:
:shipit:

jsoref commented on docs/installation/linux.md:
```diff
@@ -48,7 +58,17 @@ Then execute the following in the terminal:
 > Please note the different libicu package dependency!
```
This comment is helpful

jsoref commented on build.psm1:
```diff
@@ -1194,17 +1194,50 @@ esac
-            elseif ($IsUbuntu16) { "libicu55" }
+    # Setup package dependencies
+    # These should match those in the Dockerfiles, but exclude tools like Git, which, and curl
+    $dependencies = if ($IsUbuntu) {
```
I wish this block had a comment like the other one below.

andschwa replied:
I deduplicated it.

jsoref replied:
You de-duplicated the note about differing libicu versions?

jsoref replied:
Oh. Sorry, I just realized you pushed a newer version (perhaps based on my prodding). Yes, the newer version is much clearer, thanks!

jsoref commented on tools/download.sh:
```diff
@@ -71,23 +71,51 @@ case "$OSTYPE" in
                 ;;
             ubuntu)
                 case "$VERSION_ID" in
                     14.04)
```
I wish this had a comment like the one above

andschwa replied:
Eh this is installing literally the packages that will be listed as dependencies by the package.

jsoref replied:
I was experimenting w/ github's review feature. I think that it didn't actually make my comments easier to understand at all :-( -- The comment that I liked is the one calling the reader's attention to the differing versions of libicu.

andschwa replied:
I'm experimenting with just dpkg -i followed by apt-get install -f. It's *supposed* to work to resolve dependencies automatically. (And I don't know why I was yum installing dependencies; that's never been required.) There will probably be ugly error messages I need to hide.

jsoref reviewed (commented):
I'm not a core person to PowerShell, but looking over these commits it was hard to find the needle in the haystack -- except in the one pre-existing location where the needle was clearly identified.

If this were my project, I'd lean towards identifying the needle (as the pre-existing comment does).

You can of course take-or-leave this advice.

andschwa commented:
So, at least on Ubuntu 16.04, here are the listed depends that are categorized as "required":

* libc6
* libgcc1
* libtinfo5
* libuuid1
* zlib1g

On the one hand, these are required in the sense that .NET Core listed them as dependencies, and Ubuntu also requires them for a base system, so there is no harm in specifying them (they're already going to be installed, and PowerShell wouldn't work without them); but it's also redundant.

My *personal opinion right now* is to keep them as dependencies. I see no reason not to except to clean things up slightly (but that clean up would be technically incorrect).

Similarly, the packages

* libcurl3
* libssl1.0.0

are not *required* in the sense that PowerShell would crash without them. They are necessary for practically anything use the web cmdlets (and we have issues filed due to libcurl not being installed). Perhaps these belong in recommends or suggests instead of depends, I'm not sure. I don't see major harm in depending on them (someone, somewhere, might complain we installed something unnecessarily for their use case, maybe).

andschwa commented:
AH HA. So I've confirmed that `apt` (i.e. not `apt-get`) is indeed available by default on Ubuntu 14.04 (I tested both the ubuntu:trusty container and another developer's base install). `apt` is *much* preferably to `dpkg -i` because it (like `yum`) automatically installs dependencies for local packages (which `apt-get` cannot do). So I've updated the installation instructions and the download script to do this. Much less duplication.

`apt` is slated to replace `apt-get` suite of tools, so it is the newer version, it's just not super well known yet.

andyleejordan committed:
Add one-line installation instruction

andyleejordan committed:
Patch System.Net.Http.Native.dylib

Again, .NET Core expects users to forcibly link the third party OpenSSL
libraries into system directories, which the Homebrew team advises
strongly against (and attempts to prevent). This also affects the
System.Net.Http library, and results in runtime errors during SSL
certificate validation. So instead, we patch what we can, when we can.

andschwa commented:
@atanasa can you test that PowerShell installed via `download.sh` no longer repros #2211, and that `Start-PSBootstrap` results in a built PowerShell that also doesn't repro? I integrated your work-around into both scripts.

andschwa commented:
Got sign off from @vors, merging.
